{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 156,
   "id": "1122da3b-a989-4677-b427-4b2da5960ce8",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import networkx as nx\n",
    "\n",
    "\n",
    "\n",
    "import os\n",
    "import sys\n",
    "import inspect\n",
    "# currentdir = os.path.dirname(os.path.abspath(inspect.getfile(inspect.currentframe())))\n",
    "# parentdir = os.path.dirname(currentdir)\n",
    "# sys.path.insert(0, parentdir)\n",
    "\n",
    "sys.path.insert(0, '..')\n",
    "\n",
    "import Utils as Utils\n",
    "\n",
    "import scanpy as sc\n",
    "\n",
    "import mygene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 158,
   "id": "5938d65a-08ea-4c90-ad9e-6a0111d4918d",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 4992 × 37487\n",
       "    var: 'gene_ids', 'feature_types'"
      ]
     },
     "execution_count": 158,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_path = \"scRNA-Seq-Spatial Multiomics Reveals Metabolic Reprogramming and Calcification Characteristics of Diabetic Macroangiopathy/\";\n",
    "file_name = \"GSE248608_RAW\";\n",
    "\n",
    "adata = sc.read_10x_mtx(data_path+file_name);\n",
    "adata\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 159,
   "id": "0d37d509-d6ac-4e11-97cc-0723dd397d99",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 961 × 630\n",
       "    obs: 'n_genes'\n",
       "    var: 'gene_ids', 'feature_types', 'mito', 'n_cells'"
      ]
     },
     "execution_count": 159,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.var[\"mito\"] = adata.var_names.str.startswith(\"MT-\");\n",
    "sc.pp.filter_cells(adata, min_genes=2000);\n",
    "sc.pp.filter_genes(adata, min_cells=700);\n",
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 162,
   "id": "8297c145-484f-4d35-85eb-e27f01ece324",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Normalizing to median total counts\n",
    "sc.pp.normalize_total(adata)\n",
    "# Logarithmize the data\n",
    "sc.pp.log1p(adata)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 164,
   "id": "923a67d3-48d6-4dab-a1cf-4353a2dda52f",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 961 × 630\n",
       "    obs: 'n_genes'\n",
       "    var: 'gene_ids', 'feature_types', 'mito', 'n_cells', 'highly_variable', 'means', 'dispersions', 'dispersions_norm'\n",
       "    uns: 'log1p', 'hvg'"
      ]
     },
     "execution_count": 164,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sc.pp.highly_variable_genes(adata, n_top_genes=300, inplace=True);\n",
    "adata\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 166,
   "id": "597e8990-2506-4831-bce3-902d495c2c55",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AnnData object with n_obs × n_vars = 961 × 630\n",
      "    obs: 'n_genes'\n",
      "    var: 'gene_ids', 'feature_types', 'mito', 'n_cells', 'highly_variable', 'means', 'dispersions', 'dispersions_norm'\n",
      "    uns: 'log1p', 'hvg'\n",
      "<Compressed Sparse Column sparse matrix of dtype 'float32'\n",
      "\twith 541595 stored elements and shape (961, 630)>\n",
      "  Coords\tValues\n",
      "  (0, 0)\t1.663979172706604\n",
      "  (1, 0)\t1.2755212783813477\n",
      "  (2, 0)\t1.2161691188812256\n",
      "  (3, 0)\t1.5892910957336426\n",
      "  (4, 0)\t0.929469883441925\n",
      "  (5, 0)\t1.5196822881698608\n",
      "  (6, 0)\t1.435987949371338\n",
      "  (7, 0)\t0.571940541267395\n",
      "  (8, 0)\t1.2916241884231567\n",
      "  (9, 0)\t1.1696652173995972\n",
      "  (10, 0)\t1.3135960102081299\n",
      "  (11, 0)\t1.957932949066162\n",
      "  (12, 0)\t1.6731693744659424\n",
      "  (13, 0)\t0.8780140280723572\n",
      "  (14, 0)\t1.9093077182769775\n",
      "  (15, 0)\t1.2658189535140991\n",
      "  (16, 0)\t0.7162562608718872\n",
      "  (17, 0)\t2.0933988094329834\n",
      "  (18, 0)\t1.5407360792160034\n",
      "  (19, 0)\t0.6370176672935486\n",
      "  (20, 0)\t0.6187309622764587\n",
      "  (21, 0)\t0.7288784384727478\n",
      "  (22, 0)\t1.0599384307861328\n",
      "  (23, 0)\t1.5831726789474487\n",
      "  (24, 0)\t1.0873080492019653\n",
      "  :\t:\n",
      "  (936, 629)\t2.4723973274230957\n",
      "  (937, 629)\t2.6152420043945312\n",
      "  (938, 629)\t2.5176663398742676\n",
      "  (939, 629)\t3.377213954925537\n",
      "  (940, 629)\t3.1466853618621826\n",
      "  (941, 629)\t2.2244997024536133\n",
      "  (942, 629)\t2.4589624404907227\n",
      "  (943, 629)\t2.6569018363952637\n",
      "  (944, 629)\t1.8310116529464722\n",
      "  (945, 629)\t2.4136204719543457\n",
      "  (946, 629)\t2.389293909072876\n",
      "  (947, 629)\t2.538590908050537\n",
      "  (948, 629)\t2.5025243759155273\n",
      "  (949, 629)\t2.4881014823913574\n",
      "  (950, 629)\t2.1078226566314697\n",
      "  (951, 629)\t2.882859706878662\n",
      "  (952, 629)\t2.2516930103302\n",
      "  (953, 629)\t2.6455368995666504\n",
      "  (954, 629)\t2.2724664211273193\n",
      "  (955, 629)\t3.1131751537323\n",
      "  (956, 629)\t2.6797733306884766\n",
      "  (957, 629)\t1.8180081844329834\n",
      "  (958, 629)\t2.648061752319336\n",
      "  (959, 629)\t2.9152302742004395\n",
      "  (960, 629)\t2.767214059829712\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>n_genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACACCAATAACTGC-1</th>\n",
       "      <td>2845</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCGGAAATGTTAA-1</th>\n",
       "      <td>2091</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCGGGTAGGTACC-1</th>\n",
       "      <td>2301</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCGTTCGTCCAGG-1</th>\n",
       "      <td>2324</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACTGCTGGCTCCAA-1</th>\n",
       "      <td>3262</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTGGTCACACTCGTAA-1</th>\n",
       "      <td>3413</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTGGTTGCGGTGCGCG-1</th>\n",
       "      <td>2244</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTGTAATCCGTACTCG-1</th>\n",
       "      <td>2846</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTGTTCAGTGTGCTAC-1</th>\n",
       "      <td>2413</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTGTTTGTATTACACG-1</th>\n",
       "      <td>2700</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>961 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    n_genes\n",
       "AAACACCAATAACTGC-1     2845\n",
       "AAACCGGAAATGTTAA-1     2091\n",
       "AAACCGGGTAGGTACC-1     2301\n",
       "AAACCGTTCGTCCAGG-1     2324\n",
       "AAACTGCTGGCTCCAA-1     3262\n",
       "...                     ...\n",
       "TTGGTCACACTCGTAA-1     3413\n",
       "TTGGTTGCGGTGCGCG-1     2244\n",
       "TTGTAATCCGTACTCG-1     2846\n",
       "TTGTTCAGTGTGCTAC-1     2413\n",
       "TTGTTTGTATTACACG-1     2700\n",
       "\n",
       "[961 rows x 1 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_ids</th>\n",
       "      <th>feature_types</th>\n",
       "      <th>mito</th>\n",
       "      <th>n_cells</th>\n",
       "      <th>highly_variable</th>\n",
       "      <th>means</th>\n",
       "      <th>dispersions</th>\n",
       "      <th>dispersions_norm</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENO1</th>\n",
       "      <td>ENSG00000074800</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>False</td>\n",
       "      <td>898</td>\n",
       "      <td>False</td>\n",
       "      <td>1.402186</td>\n",
       "      <td>0.306942</td>\n",
       "      <td>-0.378558</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PARK7</th>\n",
       "      <td>ENSG00000116288</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>False</td>\n",
       "      <td>783</td>\n",
       "      <td>False</td>\n",
       "      <td>1.034365</td>\n",
       "      <td>0.163946</td>\n",
       "      <td>-0.647850</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CLIC4</th>\n",
       "      <td>ENSG00000169504</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>False</td>\n",
       "      <td>884</td>\n",
       "      <td>False</td>\n",
       "      <td>1.405928</td>\n",
       "      <td>0.313264</td>\n",
       "      <td>-0.362592</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LAPTM5</th>\n",
       "      <td>ENSG00000162511</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>False</td>\n",
       "      <td>810</td>\n",
       "      <td>True</td>\n",
       "      <td>1.289964</td>\n",
       "      <td>0.972501</td>\n",
       "      <td>2.399827</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HMGN2</th>\n",
       "      <td>ENSG00000198830</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>False</td>\n",
       "      <td>914</td>\n",
       "      <td>False</td>\n",
       "      <td>1.399874</td>\n",
       "      <td>0.207616</td>\n",
       "      <td>-0.629367</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-ND3</th>\n",
       "      <td>ENSG00000198840</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>True</td>\n",
       "      <td>961</td>\n",
       "      <td>True</td>\n",
       "      <td>2.919120</td>\n",
       "      <td>1.196903</td>\n",
       "      <td>0.502178</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-ND4L</th>\n",
       "      <td>ENSG00000212907</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>True</td>\n",
       "      <td>781</td>\n",
       "      <td>True</td>\n",
       "      <td>1.008474</td>\n",
       "      <td>0.219138</td>\n",
       "      <td>-0.025730</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-ND4</th>\n",
       "      <td>ENSG00000198886</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>True</td>\n",
       "      <td>961</td>\n",
       "      <td>True</td>\n",
       "      <td>3.104381</td>\n",
       "      <td>1.125824</td>\n",
       "      <td>0.401093</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-ND5</th>\n",
       "      <td>ENSG00000198786</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>True</td>\n",
       "      <td>802</td>\n",
       "      <td>True</td>\n",
       "      <td>1.097405</td>\n",
       "      <td>0.245785</td>\n",
       "      <td>-0.235416</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-CYB</th>\n",
       "      <td>ENSG00000198727</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>True</td>\n",
       "      <td>961</td>\n",
       "      <td>True</td>\n",
       "      <td>2.621024</td>\n",
       "      <td>0.789829</td>\n",
       "      <td>0.046533</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>630 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                gene_ids    feature_types   mito  n_cells  highly_variable  \\\n",
       "ENO1     ENSG00000074800  Gene Expression  False      898            False   \n",
       "PARK7    ENSG00000116288  Gene Expression  False      783            False   \n",
       "CLIC4    ENSG00000169504  Gene Expression  False      884            False   \n",
       "LAPTM5   ENSG00000162511  Gene Expression  False      810             True   \n",
       "HMGN2    ENSG00000198830  Gene Expression  False      914            False   \n",
       "...                  ...              ...    ...      ...              ...   \n",
       "MT-ND3   ENSG00000198840  Gene Expression   True      961             True   \n",
       "MT-ND4L  ENSG00000212907  Gene Expression   True      781             True   \n",
       "MT-ND4   ENSG00000198886  Gene Expression   True      961             True   \n",
       "MT-ND5   ENSG00000198786  Gene Expression   True      802             True   \n",
       "MT-CYB   ENSG00000198727  Gene Expression   True      961             True   \n",
       "\n",
       "            means  dispersions  dispersions_norm  \n",
       "ENO1     1.402186     0.306942         -0.378558  \n",
       "PARK7    1.034365     0.163946         -0.647850  \n",
       "CLIC4    1.405928     0.313264         -0.362592  \n",
       "LAPTM5   1.289964     0.972501          2.399827  \n",
       "HMGN2    1.399874     0.207616         -0.629367  \n",
       "...           ...          ...               ...  \n",
       "MT-ND3   2.919120     1.196903          0.502178  \n",
       "MT-ND4L  1.008474     0.219138         -0.025730  \n",
       "MT-ND4   3.104381     1.125824          0.401093  \n",
       "MT-ND5   1.097405     0.245785         -0.235416  \n",
       "MT-CYB   2.621024     0.789829          0.046533  \n",
       "\n",
       "[630 rows x 8 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "adata.var.highly_variable\n",
    "adata.X\n",
    "print(adata)\n",
    "print(adata.X)\n",
    "display(adata.obs)\n",
    "display(adata.var)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 168,
   "id": "b141c5b9-4671-410b-aad1-3e9fb37b70fc",
   "metadata": {},
   "outputs": [],
   "source": [
    "file_name = \"filtereddata\";\n",
    "\n",
    "adata.write_csvs(data_path+file_name, skip_data=False, sep=',')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 170,
   "id": "f2e2c006-72aa-4cac-b579-52c6521825a3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>n_genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACACCAATAACTGC-1</th>\n",
       "      <td>2845</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCGGAAATGTTAA-1</th>\n",
       "      <td>2091</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCGGGTAGGTACC-1</th>\n",
       "      <td>2301</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCGTTCGTCCAGG-1</th>\n",
       "      <td>2324</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACTGCTGGCTCCAA-1</th>\n",
       "      <td>3262</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTGGTCACACTCGTAA-1</th>\n",
       "      <td>3413</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTGGTTGCGGTGCGCG-1</th>\n",
       "      <td>2244</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTGTAATCCGTACTCG-1</th>\n",
       "      <td>2846</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTGTTCAGTGTGCTAC-1</th>\n",
       "      <td>2413</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTGTTTGTATTACACG-1</th>\n",
       "      <td>2700</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>961 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    n_genes\n",
       "AAACACCAATAACTGC-1     2845\n",
       "AAACCGGAAATGTTAA-1     2091\n",
       "AAACCGGGTAGGTACC-1     2301\n",
       "AAACCGTTCGTCCAGG-1     2324\n",
       "AAACTGCTGGCTCCAA-1     3262\n",
       "...                     ...\n",
       "TTGGTCACACTCGTAA-1     3413\n",
       "TTGGTTGCGGTGCGCG-1     2244\n",
       "TTGTAATCCGTACTCG-1     2846\n",
       "TTGTTCAGTGTGCTAC-1     2413\n",
       "TTGTTTGTATTACACG-1     2700\n",
       "\n",
       "[961 rows x 1 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>gene_ids</th>\n",
       "      <th>feature_types</th>\n",
       "      <th>mito</th>\n",
       "      <th>n_cells</th>\n",
       "      <th>highly_variable</th>\n",
       "      <th>means</th>\n",
       "      <th>dispersions</th>\n",
       "      <th>dispersions_norm</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENO1</td>\n",
       "      <td>ENSG00000074800</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>False</td>\n",
       "      <td>898</td>\n",
       "      <td>False</td>\n",
       "      <td>1.402186</td>\n",
       "      <td>0.306942</td>\n",
       "      <td>-0.378558</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PARK7</td>\n",
       "      <td>ENSG00000116288</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>False</td>\n",
       "      <td>783</td>\n",
       "      <td>False</td>\n",
       "      <td>1.034365</td>\n",
       "      <td>0.163946</td>\n",
       "      <td>-0.647850</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CLIC4</td>\n",
       "      <td>ENSG00000169504</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>False</td>\n",
       "      <td>884</td>\n",
       "      <td>False</td>\n",
       "      <td>1.405928</td>\n",
       "      <td>0.313264</td>\n",
       "      <td>-0.362592</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>LAPTM5</td>\n",
       "      <td>ENSG00000162511</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>False</td>\n",
       "      <td>810</td>\n",
       "      <td>True</td>\n",
       "      <td>1.289964</td>\n",
       "      <td>0.972501</td>\n",
       "      <td>2.399827</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HMGN2</td>\n",
       "      <td>ENSG00000198830</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>False</td>\n",
       "      <td>914</td>\n",
       "      <td>False</td>\n",
       "      <td>1.399874</td>\n",
       "      <td>0.207616</td>\n",
       "      <td>-0.629367</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>625</th>\n",
       "      <td>MT-ND3</td>\n",
       "      <td>ENSG00000198840</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>True</td>\n",
       "      <td>961</td>\n",
       "      <td>True</td>\n",
       "      <td>2.919120</td>\n",
       "      <td>1.196903</td>\n",
       "      <td>0.502178</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>626</th>\n",
       "      <td>MT-ND4L</td>\n",
       "      <td>ENSG00000212907</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>True</td>\n",
       "      <td>781</td>\n",
       "      <td>True</td>\n",
       "      <td>1.008474</td>\n",
       "      <td>0.219138</td>\n",
       "      <td>-0.025730</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>627</th>\n",
       "      <td>MT-ND4</td>\n",
       "      <td>ENSG00000198886</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>True</td>\n",
       "      <td>961</td>\n",
       "      <td>True</td>\n",
       "      <td>3.104381</td>\n",
       "      <td>1.125824</td>\n",
       "      <td>0.401093</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>628</th>\n",
       "      <td>MT-ND5</td>\n",
       "      <td>ENSG00000198786</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>True</td>\n",
       "      <td>802</td>\n",
       "      <td>True</td>\n",
       "      <td>1.097405</td>\n",
       "      <td>0.245785</td>\n",
       "      <td>-0.235416</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>629</th>\n",
       "      <td>MT-CYB</td>\n",
       "      <td>ENSG00000198727</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>True</td>\n",
       "      <td>961</td>\n",
       "      <td>True</td>\n",
       "      <td>2.621024</td>\n",
       "      <td>0.789829</td>\n",
       "      <td>0.046533</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>630 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Unnamed: 0         gene_ids    feature_types   mito  n_cells  \\\n",
       "0         ENO1  ENSG00000074800  Gene Expression  False      898   \n",
       "1        PARK7  ENSG00000116288  Gene Expression  False      783   \n",
       "2        CLIC4  ENSG00000169504  Gene Expression  False      884   \n",
       "3       LAPTM5  ENSG00000162511  Gene Expression  False      810   \n",
       "4        HMGN2  ENSG00000198830  Gene Expression  False      914   \n",
       "..         ...              ...              ...    ...      ...   \n",
       "625     MT-ND3  ENSG00000198840  Gene Expression   True      961   \n",
       "626    MT-ND4L  ENSG00000212907  Gene Expression   True      781   \n",
       "627     MT-ND4  ENSG00000198886  Gene Expression   True      961   \n",
       "628     MT-ND5  ENSG00000198786  Gene Expression   True      802   \n",
       "629     MT-CYB  ENSG00000198727  Gene Expression   True      961   \n",
       "\n",
       "     highly_variable     means  dispersions  dispersions_norm  \n",
       "0              False  1.402186     0.306942         -0.378558  \n",
       "1              False  1.034365     0.163946         -0.647850  \n",
       "2              False  1.405928     0.313264         -0.362592  \n",
       "3               True  1.289964     0.972501          2.399827  \n",
       "4              False  1.399874     0.207616         -0.629367  \n",
       "..               ...       ...          ...               ...  \n",
       "625             True  2.919120     1.196903          0.502178  \n",
       "626             True  1.008474     0.219138         -0.025730  \n",
       "627             True  3.104381     1.125824          0.401093  \n",
       "628             True  1.097405     0.245785         -0.235416  \n",
       "629             True  2.621024     0.789829          0.046533  \n",
       "\n",
       "[630 rows x 9 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ENSG00000074800</th>\n",
       "      <th>ENSG00000116288</th>\n",
       "      <th>ENSG00000169504</th>\n",
       "      <th>ENSG00000162511</th>\n",
       "      <th>ENSG00000198830</th>\n",
       "      <th>ENSG00000116251</th>\n",
       "      <th>ENSG00000077549</th>\n",
       "      <th>ENSG00000117519</th>\n",
       "      <th>ENSG00000159377</th>\n",
       "      <th>ENSG00000116473</th>\n",
       "      <th>...</th>\n",
       "      <th>ENSG00000198763</th>\n",
       "      <th>ENSG00000198804</th>\n",
       "      <th>ENSG00000198712</th>\n",
       "      <th>ENSG00000198899</th>\n",
       "      <th>ENSG00000198938</th>\n",
       "      <th>ENSG00000198840</th>\n",
       "      <th>ENSG00000212907</th>\n",
       "      <th>ENSG00000198886</th>\n",
       "      <th>ENSG00000198786</th>\n",
       "      <th>ENSG00000198727</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACACCAATAACTGC-1</th>\n",
       "      <td>1.663979</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.997728</td>\n",
       "      <td>1.814228</td>\n",
       "      <td>0.618454</td>\n",
       "      <td>2.257646</td>\n",
       "      <td>0.997728</td>\n",
       "      <td>1.814228</td>\n",
       "      <td>1.272052</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>2.422381</td>\n",
       "      <td>2.495577</td>\n",
       "      <td>2.687636</td>\n",
       "      <td>2.060316</td>\n",
       "      <td>2.422381</td>\n",
       "      <td>2.563778</td>\n",
       "      <td>0.997728</td>\n",
       "      <td>2.563778</td>\n",
       "      <td>0.997728</td>\n",
       "      <td>2.163841</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCGGAAATGTTAA-1</th>\n",
       "      <td>1.275521</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.275521</td>\n",
       "      <td>0.828676</td>\n",
       "      <td>2.008404</td>\n",
       "      <td>2.305779</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.828676</td>\n",
       "      <td>...</td>\n",
       "      <td>2.305779</td>\n",
       "      <td>2.534692</td>\n",
       "      <td>2.632093</td>\n",
       "      <td>1.583269</td>\n",
       "      <td>2.534692</td>\n",
       "      <td>2.168105</td>\n",
       "      <td>0.828676</td>\n",
       "      <td>2.632093</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.168105</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCGGGTAGGTACC-1</th>\n",
       "      <td>1.216169</td>\n",
       "      <td>1.216169</td>\n",
       "      <td>0.782585</td>\n",
       "      <td>1.216169</td>\n",
       "      <td>0.782585</td>\n",
       "      <td>1.936666</td>\n",
       "      <td>1.936666</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.782585</td>\n",
       "      <td>0.782585</td>\n",
       "      <td>...</td>\n",
       "      <td>1.517620</td>\n",
       "      <td>2.869015</td>\n",
       "      <td>3.255381</td>\n",
       "      <td>2.554991</td>\n",
       "      <td>3.107637</td>\n",
       "      <td>3.384080</td>\n",
       "      <td>1.517620</td>\n",
       "      <td>3.300146</td>\n",
       "      <td>1.748934</td>\n",
       "      <td>2.643213</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCGTTCGTCCAGG-1</th>\n",
       "      <td>1.589291</td>\n",
       "      <td>1.589291</td>\n",
       "      <td>0.832949</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.014964</td>\n",
       "      <td>2.014964</td>\n",
       "      <td>0.832949</td>\n",
       "      <td>0.832949</td>\n",
       "      <td>1.280984</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>2.433677</td>\n",
       "      <td>2.541667</td>\n",
       "      <td>2.954977</td>\n",
       "      <td>2.174814</td>\n",
       "      <td>3.342930</td>\n",
       "      <td>2.954977</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.809469</td>\n",
       "      <td>0.832949</td>\n",
       "      <td>2.727918</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACTGCTGGCTCCAA-1</th>\n",
       "      <td>0.929470</td>\n",
       "      <td>0.569047</td>\n",
       "      <td>0.929470</td>\n",
       "      <td>0.929470</td>\n",
       "      <td>1.575450</td>\n",
       "      <td>1.964685</td>\n",
       "      <td>1.402741</td>\n",
       "      <td>0.929470</td>\n",
       "      <td>0.929470</td>\n",
       "      <td>1.193846</td>\n",
       "      <td>...</td>\n",
       "      <td>2.694526</td>\n",
       "      <td>2.838976</td>\n",
       "      <td>3.353301</td>\n",
       "      <td>3.209394</td>\n",
       "      <td>3.145484</td>\n",
       "      <td>3.298198</td>\n",
       "      <td>1.402741</td>\n",
       "      <td>3.145484</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.694526</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTGGTCACACTCGTAA-1</th>\n",
       "      <td>0.562209</td>\n",
       "      <td>1.390828</td>\n",
       "      <td>1.390828</td>\n",
       "      <td>0.562209</td>\n",
       "      <td>2.052956</td>\n",
       "      <td>1.390828</td>\n",
       "      <td>0.919919</td>\n",
       "      <td>0.919919</td>\n",
       "      <td>1.182841</td>\n",
       "      <td>1.390828</td>\n",
       "      <td>...</td>\n",
       "      <td>2.052956</td>\n",
       "      <td>2.308023</td>\n",
       "      <td>3.194208</td>\n",
       "      <td>2.447863</td>\n",
       "      <td>3.062114</td>\n",
       "      <td>2.447863</td>\n",
       "      <td>0.919919</td>\n",
       "      <td>3.130341</td>\n",
       "      <td>0.562209</td>\n",
       "      <td>2.679773</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTGGTTGCGGTGCGCG-1</th>\n",
       "      <td>1.275304</td>\n",
       "      <td>0.828506</td>\n",
       "      <td>1.275304</td>\n",
       "      <td>1.275304</td>\n",
       "      <td>2.008143</td>\n",
       "      <td>1.583029</td>\n",
       "      <td>1.275304</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>2.426496</td>\n",
       "      <td>2.534415</td>\n",
       "      <td>3.074466</td>\n",
       "      <td>1.275304</td>\n",
       "      <td>2.426496</td>\n",
       "      <td>2.426496</td>\n",
       "      <td>0.828506</td>\n",
       "      <td>2.720561</td>\n",
       "      <td>1.583029</td>\n",
       "      <td>1.818008</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTGTAATCCGTACTCG-1</th>\n",
       "      <td>1.908490</td>\n",
       "      <td>1.454340</td>\n",
       "      <td>1.454340</td>\n",
       "      <td>0.971093</td>\n",
       "      <td>0.971093</td>\n",
       "      <td>2.383702</td>\n",
       "      <td>0.971093</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.599064</td>\n",
       "      <td>1.241628</td>\n",
       "      <td>...</td>\n",
       "      <td>2.126293</td>\n",
       "      <td>2.648062</td>\n",
       "      <td>2.648062</td>\n",
       "      <td>3.142046</td>\n",
       "      <td>2.704514</td>\n",
       "      <td>3.176870</td>\n",
       "      <td>1.778756</td>\n",
       "      <td>2.902998</td>\n",
       "      <td>1.778756</td>\n",
       "      <td>2.648062</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTGTTCAGTGTGCTAC-1</th>\n",
       "      <td>1.250845</td>\n",
       "      <td>1.250845</td>\n",
       "      <td>1.789522</td>\n",
       "      <td>1.556025</td>\n",
       "      <td>0.809439</td>\n",
       "      <td>1.789522</td>\n",
       "      <td>1.556025</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.809439</td>\n",
       "      <td>0.809439</td>\n",
       "      <td>...</td>\n",
       "      <td>2.600203</td>\n",
       "      <td>2.688740</td>\n",
       "      <td>3.206248</td>\n",
       "      <td>2.600203</td>\n",
       "      <td>2.770071</td>\n",
       "      <td>3.206248</td>\n",
       "      <td>1.250845</td>\n",
       "      <td>3.302466</td>\n",
       "      <td>1.250845</td>\n",
       "      <td>2.915230</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTGTTTGTATTACACG-1</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.013278</td>\n",
       "      <td>1.683903</td>\n",
       "      <td>1.289763</td>\n",
       "      <td>0.629839</td>\n",
       "      <td>1.013278</td>\n",
       "      <td>1.013278</td>\n",
       "      <td>1.683903</td>\n",
       "      <td>0.629839</td>\n",
       "      <td>0.629839</td>\n",
       "      <td>...</td>\n",
       "      <td>2.586433</td>\n",
       "      <td>2.710509</td>\n",
       "      <td>3.339200</td>\n",
       "      <td>3.132564</td>\n",
       "      <td>3.510363</td>\n",
       "      <td>3.428439</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.483793</td>\n",
       "      <td>1.289763</td>\n",
       "      <td>2.767214</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>961 rows × 630 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    ENSG00000074800  ENSG00000116288  ENSG00000169504  \\\n",
       "AAACACCAATAACTGC-1         1.663979         0.000000         0.997728   \n",
       "AAACCGGAAATGTTAA-1         1.275521         0.000000         1.275521   \n",
       "AAACCGGGTAGGTACC-1         1.216169         1.216169         0.782585   \n",
       "AAACCGTTCGTCCAGG-1         1.589291         1.589291         0.832949   \n",
       "AAACTGCTGGCTCCAA-1         0.929470         0.569047         0.929470   \n",
       "...                             ...              ...              ...   \n",
       "TTGGTCACACTCGTAA-1         0.562209         1.390828         1.390828   \n",
       "TTGGTTGCGGTGCGCG-1         1.275304         0.828506         1.275304   \n",
       "TTGTAATCCGTACTCG-1         1.908490         1.454340         1.454340   \n",
       "TTGTTCAGTGTGCTAC-1         1.250845         1.250845         1.789522   \n",
       "TTGTTTGTATTACACG-1         0.000000         1.013278         1.683903   \n",
       "\n",
       "                    ENSG00000162511  ENSG00000198830  ENSG00000116251  \\\n",
       "AAACACCAATAACTGC-1         1.814228         0.618454         2.257646   \n",
       "AAACCGGAAATGTTAA-1         0.828676         2.008404         2.305779   \n",
       "AAACCGGGTAGGTACC-1         1.216169         0.782585         1.936666   \n",
       "AAACCGTTCGTCCAGG-1         0.000000         2.014964         2.014964   \n",
       "AAACTGCTGGCTCCAA-1         0.929470         1.575450         1.964685   \n",
       "...                             ...              ...              ...   \n",
       "TTGGTCACACTCGTAA-1         0.562209         2.052956         1.390828   \n",
       "TTGGTTGCGGTGCGCG-1         1.275304         2.008143         1.583029   \n",
       "TTGTAATCCGTACTCG-1         0.971093         0.971093         2.383702   \n",
       "TTGTTCAGTGTGCTAC-1         1.556025         0.809439         1.789522   \n",
       "TTGTTTGTATTACACG-1         1.289763         0.629839         1.013278   \n",
       "\n",
       "                    ENSG00000077549  ENSG00000117519  ENSG00000159377  \\\n",
       "AAACACCAATAACTGC-1         0.997728         1.814228         1.272052   \n",
       "AAACCGGAAATGTTAA-1         0.000000         0.000000         0.000000   \n",
       "AAACCGGGTAGGTACC-1         1.936666         0.000000         0.782585   \n",
       "AAACCGTTCGTCCAGG-1         0.832949         0.832949         1.280984   \n",
       "AAACTGCTGGCTCCAA-1         1.402741         0.929470         0.929470   \n",
       "...                             ...              ...              ...   \n",
       "TTGGTCACACTCGTAA-1         0.919919         0.919919         1.182841   \n",
       "TTGGTTGCGGTGCGCG-1         1.275304         0.000000         0.000000   \n",
       "TTGTAATCCGTACTCG-1         0.971093         0.000000         0.599064   \n",
       "TTGTTCAGTGTGCTAC-1         1.556025         0.000000         0.809439   \n",
       "TTGTTTGTATTACACG-1         1.013278         1.683903         0.629839   \n",
       "\n",
       "                    ENSG00000116473  ...  ENSG00000198763  ENSG00000198804  \\\n",
       "AAACACCAATAACTGC-1         0.000000  ...         2.422381         2.495577   \n",
       "AAACCGGAAATGTTAA-1         0.828676  ...         2.305779         2.534692   \n",
       "AAACCGGGTAGGTACC-1         0.782585  ...         1.517620         2.869015   \n",
       "AAACCGTTCGTCCAGG-1         0.000000  ...         2.433677         2.541667   \n",
       "AAACTGCTGGCTCCAA-1         1.193846  ...         2.694526         2.838976   \n",
       "...                             ...  ...              ...              ...   \n",
       "TTGGTCACACTCGTAA-1         1.390828  ...         2.052956         2.308023   \n",
       "TTGGTTGCGGTGCGCG-1         0.000000  ...         2.426496         2.534415   \n",
       "TTGTAATCCGTACTCG-1         1.241628  ...         2.126293         2.648062   \n",
       "TTGTTCAGTGTGCTAC-1         0.809439  ...         2.600203         2.688740   \n",
       "TTGTTTGTATTACACG-1         0.629839  ...         2.586433         2.710509   \n",
       "\n",
       "                    ENSG00000198712  ENSG00000198899  ENSG00000198938  \\\n",
       "AAACACCAATAACTGC-1         2.687636         2.060316         2.422381   \n",
       "AAACCGGAAATGTTAA-1         2.632093         1.583269         2.534692   \n",
       "AAACCGGGTAGGTACC-1         3.255381         2.554991         3.107637   \n",
       "AAACCGTTCGTCCAGG-1         2.954977         2.174814         3.342930   \n",
       "AAACTGCTGGCTCCAA-1         3.353301         3.209394         3.145484   \n",
       "...                             ...              ...              ...   \n",
       "TTGGTCACACTCGTAA-1         3.194208         2.447863         3.062114   \n",
       "TTGGTTGCGGTGCGCG-1         3.074466         1.275304         2.426496   \n",
       "TTGTAATCCGTACTCG-1         2.648062         3.142046         2.704514   \n",
       "TTGTTCAGTGTGCTAC-1         3.206248         2.600203         2.770071   \n",
       "TTGTTTGTATTACACG-1         3.339200         3.132564         3.510363   \n",
       "\n",
       "                    ENSG00000198840  ENSG00000212907  ENSG00000198886  \\\n",
       "AAACACCAATAACTGC-1         2.563778         0.997728         2.563778   \n",
       "AAACCGGAAATGTTAA-1         2.168105         0.828676         2.632093   \n",
       "AAACCGGGTAGGTACC-1         3.384080         1.517620         3.300146   \n",
       "AAACCGTTCGTCCAGG-1         2.954977         0.000000         2.809469   \n",
       "AAACTGCTGGCTCCAA-1         3.298198         1.402741         3.145484   \n",
       "...                             ...              ...              ...   \n",
       "TTGGTCACACTCGTAA-1         2.447863         0.919919         3.130341   \n",
       "TTGGTTGCGGTGCGCG-1         2.426496         0.828506         2.720561   \n",
       "TTGTAATCCGTACTCG-1         3.176870         1.778756         2.902998   \n",
       "TTGTTCAGTGTGCTAC-1         3.206248         1.250845         3.302466   \n",
       "TTGTTTGTATTACACG-1         3.428439         0.000000         3.483793   \n",
       "\n",
       "                    ENSG00000198786  ENSG00000198727  \n",
       "AAACACCAATAACTGC-1         0.997728         2.163841  \n",
       "AAACCGGAAATGTTAA-1         0.000000         2.168105  \n",
       "AAACCGGGTAGGTACC-1         1.748934         2.643213  \n",
       "AAACCGTTCGTCCAGG-1         0.832949         2.727918  \n",
       "AAACTGCTGGCTCCAA-1         0.000000         2.694526  \n",
       "...                             ...              ...  \n",
       "TTGGTCACACTCGTAA-1         0.562209         2.679773  \n",
       "TTGGTTGCGGTGCGCG-1         1.583029         1.818008  \n",
       "TTGTAATCCGTACTCG-1         1.778756         2.648062  \n",
       "TTGTTCAGTGTGCTAC-1         1.250845         2.915230  \n",
       "TTGTTTGTATTACACG-1         1.289763         2.767214  \n",
       "\n",
       "[961 rows x 630 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ENSG00000162511</th>\n",
       "      <th>ENSG00000117519</th>\n",
       "      <th>ENSG00000159377</th>\n",
       "      <th>ENSG00000116473</th>\n",
       "      <th>ENSG00000163191</th>\n",
       "      <th>ENSG00000000971</th>\n",
       "      <th>ENSG00000265972</th>\n",
       "      <th>ENSG00000159176</th>\n",
       "      <th>ENSG00000143248</th>\n",
       "      <th>ENSG00000077157</th>\n",
       "      <th>...</th>\n",
       "      <th>ENSG00000198763</th>\n",
       "      <th>ENSG00000198804</th>\n",
       "      <th>ENSG00000198712</th>\n",
       "      <th>ENSG00000198899</th>\n",
       "      <th>ENSG00000198938</th>\n",
       "      <th>ENSG00000198840</th>\n",
       "      <th>ENSG00000212907</th>\n",
       "      <th>ENSG00000198886</th>\n",
       "      <th>ENSG00000198786</th>\n",
       "      <th>ENSG00000198727</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACACCAATAACTGC-1</th>\n",
       "      <td>1.814228</td>\n",
       "      <td>1.814228</td>\n",
       "      <td>1.272052</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.944823</td>\n",
       "      <td>1.487095</td>\n",
       "      <td>1.663979</td>\n",
       "      <td>3.146626</td>\n",
       "      <td>1.814228</td>\n",
       "      <td>0.618454</td>\n",
       "      <td>...</td>\n",
       "      <td>2.422381</td>\n",
       "      <td>2.495577</td>\n",
       "      <td>2.687636</td>\n",
       "      <td>2.060316</td>\n",
       "      <td>2.422381</td>\n",
       "      <td>2.563778</td>\n",
       "      <td>0.997728</td>\n",
       "      <td>2.563778</td>\n",
       "      <td>0.997728</td>\n",
       "      <td>2.163841</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCGGAAATGTTAA-1</th>\n",
       "      <td>0.828676</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.828676</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.275521</td>\n",
       "      <td>0.828676</td>\n",
       "      <td>3.380527</td>\n",
       "      <td>2.534692</td>\n",
       "      <td>1.275521</td>\n",
       "      <td>...</td>\n",
       "      <td>2.305779</td>\n",
       "      <td>2.534692</td>\n",
       "      <td>2.632093</td>\n",
       "      <td>1.583269</td>\n",
       "      <td>2.534692</td>\n",
       "      <td>2.168105</td>\n",
       "      <td>0.828676</td>\n",
       "      <td>2.632093</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.168105</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCGGGTAGGTACC-1</th>\n",
       "      <td>1.216169</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.782585</td>\n",
       "      <td>0.782585</td>\n",
       "      <td>1.748934</td>\n",
       "      <td>1.748934</td>\n",
       "      <td>1.748934</td>\n",
       "      <td>2.934217</td>\n",
       "      <td>2.869015</td>\n",
       "      <td>1.748934</td>\n",
       "      <td>...</td>\n",
       "      <td>1.517620</td>\n",
       "      <td>2.869015</td>\n",
       "      <td>3.255381</td>\n",
       "      <td>2.554991</td>\n",
       "      <td>3.107637</td>\n",
       "      <td>3.384080</td>\n",
       "      <td>1.517620</td>\n",
       "      <td>3.300146</td>\n",
       "      <td>1.748934</td>\n",
       "      <td>2.643213</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCGTTCGTCCAGG-1</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.832949</td>\n",
       "      <td>1.280984</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.280984</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.832949</td>\n",
       "      <td>3.549685</td>\n",
       "      <td>2.884867</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>2.433677</td>\n",
       "      <td>2.541667</td>\n",
       "      <td>2.954977</td>\n",
       "      <td>2.174814</td>\n",
       "      <td>3.342930</td>\n",
       "      <td>2.954977</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.809469</td>\n",
       "      <td>0.832949</td>\n",
       "      <td>2.727918</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACTGCTGGCTCCAA-1</th>\n",
       "      <td>0.929470</td>\n",
       "      <td>0.929470</td>\n",
       "      <td>0.929470</td>\n",
       "      <td>1.193846</td>\n",
       "      <td>1.193846</td>\n",
       "      <td>1.193846</td>\n",
       "      <td>1.402741</td>\n",
       "      <td>3.209394</td>\n",
       "      <td>2.159388</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>2.694526</td>\n",
       "      <td>2.838976</td>\n",
       "      <td>3.353301</td>\n",
       "      <td>3.209394</td>\n",
       "      <td>3.145484</td>\n",
       "      <td>3.298198</td>\n",
       "      <td>1.402741</td>\n",
       "      <td>3.145484</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.694526</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTGGTCACACTCGTAA-1</th>\n",
       "      <td>0.562209</td>\n",
       "      <td>0.919919</td>\n",
       "      <td>1.182841</td>\n",
       "      <td>1.390828</td>\n",
       "      <td>1.562916</td>\n",
       "      <td>1.562916</td>\n",
       "      <td>0.919919</td>\n",
       "      <td>3.509914</td>\n",
       "      <td>2.950162</td>\n",
       "      <td>1.390828</td>\n",
       "      <td>...</td>\n",
       "      <td>2.052956</td>\n",
       "      <td>2.308023</td>\n",
       "      <td>3.194208</td>\n",
       "      <td>2.447863</td>\n",
       "      <td>3.062114</td>\n",
       "      <td>2.447863</td>\n",
       "      <td>0.919919</td>\n",
       "      <td>3.130341</td>\n",
       "      <td>0.562209</td>\n",
       "      <td>2.679773</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTGGTTGCGGTGCGCG-1</th>\n",
       "      <td>1.275304</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.583029</td>\n",
       "      <td>1.275304</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.802071</td>\n",
       "      <td>1.818008</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>2.426496</td>\n",
       "      <td>2.534415</td>\n",
       "      <td>3.074466</td>\n",
       "      <td>1.275304</td>\n",
       "      <td>2.426496</td>\n",
       "      <td>2.426496</td>\n",
       "      <td>0.828506</td>\n",
       "      <td>2.720561</td>\n",
       "      <td>1.583029</td>\n",
       "      <td>1.818008</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTGTAATCCGTACTCG-1</th>\n",
       "      <td>0.971093</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.599064</td>\n",
       "      <td>1.241628</td>\n",
       "      <td>1.241628</td>\n",
       "      <td>2.023310</td>\n",
       "      <td>0.971093</td>\n",
       "      <td>3.142046</td>\n",
       "      <td>3.029648</td>\n",
       "      <td>0.599064</td>\n",
       "      <td>...</td>\n",
       "      <td>2.126293</td>\n",
       "      <td>2.648062</td>\n",
       "      <td>2.648062</td>\n",
       "      <td>3.142046</td>\n",
       "      <td>2.704514</td>\n",
       "      <td>3.176870</td>\n",
       "      <td>1.778756</td>\n",
       "      <td>2.902998</td>\n",
       "      <td>1.778756</td>\n",
       "      <td>2.648062</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTGTTCAGTGTGCTAC-1</th>\n",
       "      <td>1.556025</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.809439</td>\n",
       "      <td>0.809439</td>\n",
       "      <td>1.556025</td>\n",
       "      <td>1.556025</td>\n",
       "      <td>1.250845</td>\n",
       "      <td>2.980604</td>\n",
       "      <td>2.845282</td>\n",
       "      <td>0.809439</td>\n",
       "      <td>...</td>\n",
       "      <td>2.600203</td>\n",
       "      <td>2.688740</td>\n",
       "      <td>3.206248</td>\n",
       "      <td>2.600203</td>\n",
       "      <td>2.770071</td>\n",
       "      <td>3.206248</td>\n",
       "      <td>1.250845</td>\n",
       "      <td>3.302466</td>\n",
       "      <td>1.250845</td>\n",
       "      <td>2.915230</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTGTTTGTATTACACG-1</th>\n",
       "      <td>1.289763</td>\n",
       "      <td>1.683903</td>\n",
       "      <td>0.629839</td>\n",
       "      <td>0.629839</td>\n",
       "      <td>1.506127</td>\n",
       "      <td>1.683903</td>\n",
       "      <td>1.506127</td>\n",
       "      <td>2.920263</td>\n",
       "      <td>2.081747</td>\n",
       "      <td>1.506127</td>\n",
       "      <td>...</td>\n",
       "      <td>2.586433</td>\n",
       "      <td>2.710509</td>\n",
       "      <td>3.339200</td>\n",
       "      <td>3.132564</td>\n",
       "      <td>3.510363</td>\n",
       "      <td>3.428439</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.483793</td>\n",
       "      <td>1.289763</td>\n",
       "      <td>2.767214</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>961 rows × 300 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    ENSG00000162511  ENSG00000117519  ENSG00000159377  \\\n",
       "AAACACCAATAACTGC-1         1.814228         1.814228         1.272052   \n",
       "AAACCGGAAATGTTAA-1         0.828676         0.000000         0.000000   \n",
       "AAACCGGGTAGGTACC-1         1.216169         0.000000         0.782585   \n",
       "AAACCGTTCGTCCAGG-1         0.000000         0.832949         1.280984   \n",
       "AAACTGCTGGCTCCAA-1         0.929470         0.929470         0.929470   \n",
       "...                             ...              ...              ...   \n",
       "TTGGTCACACTCGTAA-1         0.562209         0.919919         1.182841   \n",
       "TTGGTTGCGGTGCGCG-1         1.275304         0.000000         0.000000   \n",
       "TTGTAATCCGTACTCG-1         0.971093         0.000000         0.599064   \n",
       "TTGTTCAGTGTGCTAC-1         1.556025         0.000000         0.809439   \n",
       "TTGTTTGTATTACACG-1         1.289763         1.683903         0.629839   \n",
       "\n",
       "                    ENSG00000116473  ENSG00000163191  ENSG00000000971  \\\n",
       "AAACACCAATAACTGC-1         0.000000         1.944823         1.487095   \n",
       "AAACCGGAAATGTTAA-1         0.828676         0.000000         1.275521   \n",
       "AAACCGGGTAGGTACC-1         0.782585         1.748934         1.748934   \n",
       "AAACCGTTCGTCCAGG-1         0.000000         1.280984         0.000000   \n",
       "AAACTGCTGGCTCCAA-1         1.193846         1.193846         1.193846   \n",
       "...                             ...              ...              ...   \n",
       "TTGGTCACACTCGTAA-1         1.390828         1.562916         1.562916   \n",
       "TTGGTTGCGGTGCGCG-1         0.000000         1.583029         1.275304   \n",
       "TTGTAATCCGTACTCG-1         1.241628         1.241628         2.023310   \n",
       "TTGTTCAGTGTGCTAC-1         0.809439         1.556025         1.556025   \n",
       "TTGTTTGTATTACACG-1         0.629839         1.506127         1.683903   \n",
       "\n",
       "                    ENSG00000265972  ENSG00000159176  ENSG00000143248  \\\n",
       "AAACACCAATAACTGC-1         1.663979         3.146626         1.814228   \n",
       "AAACCGGAAATGTTAA-1         0.828676         3.380527         2.534692   \n",
       "AAACCGGGTAGGTACC-1         1.748934         2.934217         2.869015   \n",
       "AAACCGTTCGTCCAGG-1         0.832949         3.549685         2.884867   \n",
       "AAACTGCTGGCTCCAA-1         1.402741         3.209394         2.159388   \n",
       "...                             ...              ...              ...   \n",
       "TTGGTCACACTCGTAA-1         0.919919         3.509914         2.950162   \n",
       "TTGGTTGCGGTGCGCG-1         0.000000         2.802071         1.818008   \n",
       "TTGTAATCCGTACTCG-1         0.971093         3.142046         3.029648   \n",
       "TTGTTCAGTGTGCTAC-1         1.250845         2.980604         2.845282   \n",
       "TTGTTTGTATTACACG-1         1.506127         2.920263         2.081747   \n",
       "\n",
       "                    ENSG00000077157  ...  ENSG00000198763  ENSG00000198804  \\\n",
       "AAACACCAATAACTGC-1         0.618454  ...         2.422381         2.495577   \n",
       "AAACCGGAAATGTTAA-1         1.275521  ...         2.305779         2.534692   \n",
       "AAACCGGGTAGGTACC-1         1.748934  ...         1.517620         2.869015   \n",
       "AAACCGTTCGTCCAGG-1         0.000000  ...         2.433677         2.541667   \n",
       "AAACTGCTGGCTCCAA-1         0.000000  ...         2.694526         2.838976   \n",
       "...                             ...  ...              ...              ...   \n",
       "TTGGTCACACTCGTAA-1         1.390828  ...         2.052956         2.308023   \n",
       "TTGGTTGCGGTGCGCG-1         0.000000  ...         2.426496         2.534415   \n",
       "TTGTAATCCGTACTCG-1         0.599064  ...         2.126293         2.648062   \n",
       "TTGTTCAGTGTGCTAC-1         0.809439  ...         2.600203         2.688740   \n",
       "TTGTTTGTATTACACG-1         1.506127  ...         2.586433         2.710509   \n",
       "\n",
       "                    ENSG00000198712  ENSG00000198899  ENSG00000198938  \\\n",
       "AAACACCAATAACTGC-1         2.687636         2.060316         2.422381   \n",
       "AAACCGGAAATGTTAA-1         2.632093         1.583269         2.534692   \n",
       "AAACCGGGTAGGTACC-1         3.255381         2.554991         3.107637   \n",
       "AAACCGTTCGTCCAGG-1         2.954977         2.174814         3.342930   \n",
       "AAACTGCTGGCTCCAA-1         3.353301         3.209394         3.145484   \n",
       "...                             ...              ...              ...   \n",
       "TTGGTCACACTCGTAA-1         3.194208         2.447863         3.062114   \n",
       "TTGGTTGCGGTGCGCG-1         3.074466         1.275304         2.426496   \n",
       "TTGTAATCCGTACTCG-1         2.648062         3.142046         2.704514   \n",
       "TTGTTCAGTGTGCTAC-1         3.206248         2.600203         2.770071   \n",
       "TTGTTTGTATTACACG-1         3.339200         3.132564         3.510363   \n",
       "\n",
       "                    ENSG00000198840  ENSG00000212907  ENSG00000198886  \\\n",
       "AAACACCAATAACTGC-1         2.563778         0.997728         2.563778   \n",
       "AAACCGGAAATGTTAA-1         2.168105         0.828676         2.632093   \n",
       "AAACCGGGTAGGTACC-1         3.384080         1.517620         3.300146   \n",
       "AAACCGTTCGTCCAGG-1         2.954977         0.000000         2.809469   \n",
       "AAACTGCTGGCTCCAA-1         3.298198         1.402741         3.145484   \n",
       "...                             ...              ...              ...   \n",
       "TTGGTCACACTCGTAA-1         2.447863         0.919919         3.130341   \n",
       "TTGGTTGCGGTGCGCG-1         2.426496         0.828506         2.720561   \n",
       "TTGTAATCCGTACTCG-1         3.176870         1.778756         2.902998   \n",
       "TTGTTCAGTGTGCTAC-1         3.206248         1.250845         3.302466   \n",
       "TTGTTTGTATTACACG-1         3.428439         0.000000         3.483793   \n",
       "\n",
       "                    ENSG00000198786  ENSG00000198727  \n",
       "AAACACCAATAACTGC-1         0.997728         2.163841  \n",
       "AAACCGGAAATGTTAA-1         0.000000         2.168105  \n",
       "AAACCGGGTAGGTACC-1         1.748934         2.643213  \n",
       "AAACCGTTCGTCCAGG-1         0.832949         2.727918  \n",
       "AAACTGCTGGCTCCAA-1         0.000000         2.694526  \n",
       "...                             ...              ...  \n",
       "TTGGTCACACTCGTAA-1         0.562209         2.679773  \n",
       "TTGGTTGCGGTGCGCG-1         1.583029         1.818008  \n",
       "TTGTAATCCGTACTCG-1         1.778756         2.648062  \n",
       "TTGTTCAGTGTGCTAC-1         1.250845         2.915230  \n",
       "TTGTTTGTATTACACG-1         1.289763         2.767214  \n",
       "\n",
       "[961 rows x 300 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "obs = pd.read_csv(data_path+file_name+\"/obs.csv\",index_col=0);\n",
    "display(obs);\n",
    "var = pd.read_csv(data_path+file_name+\"/var.csv\");\n",
    "display(var);\n",
    "\n",
    "dataset = pd.read_csv(data_path+file_name+\"/X.csv\",names=var['gene_ids']);\n",
    "dataset.index = obs.index;\n",
    "display(dataset);\n",
    "\n",
    "highly_variable_genes = var[var['highly_variable']]['gene_ids'].values;\n",
    "dataset = dataset[highly_variable_genes];\n",
    "display(dataset);"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 172,
   "id": "e6c9181a-e6ad-47d5-a82b-bb63b2ec5673",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "1 input query terms found no hit:\t['ENSG00000112096']\n"
     ]
    }
   ],
   "source": [
    "gene_list = dataset.columns.tolist();\n",
    "\n",
    "mg = mygene.MyGeneInfo();\n",
    "\n",
    "geneSyms = mg.querymany(gene_list, scopes='ensembl.gene', fields='symbol', species='human');\n",
    "\n",
    "gene_info_dict = {};\n",
    "for ele in geneSyms:\n",
    "    # print(ele);\n",
    "    query = ele['query'];\n",
    "    if 'symbol' in ele:\n",
    "        symbol = ele['symbol'];\n",
    "        gene_info_dict[query] = symbol;\n",
    "    \n",
    "dataset.rename(columns=gene_info_dict, inplace=True);\n",
    "dataset\n",
    "dataset.to_csv(data_path+file_name+\"/processed_data.csv\",index=None);"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fb91f312-114a-4017-999b-2bf08f07cce1",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
